### ISSUE TITLE
Icon: split background/span icons from foreground/svg icons

### ISSUE THREAD

**Issue by ianmcburnie** [eBay/skin#239]
The icon.css currently contains the rules for both background and foreground icons.

For example:

```css
svg.icon--arrow-left {
  height: 19px;
  width: 21px;
}
span.icon--arrow-left {
  background-image: url('data:image/svg+xml;base64,--snip--');
  height: 19px;
  width: 21px;
}
svg.icon--arrow-right {
  height: 19px;
  width: 21px;
}
span.icon--arrow-right {
  background-image: url('data:image/svg+xml;base64,--snip--');
  height: 19px;
  width: 21px;
}
```

This results in redundancy for apps that only use inline icons and no background icons (or vice versa). This redundancy is going to significantly increase when we add high contrast versions of background icons (see #233).

Let's move these rules to separate dist files and create specific modules (e.g. `@ebay/skin/icon/bg` & `@ebay/skin/icon/fg`). If we want to maintain backwards compatibility, the current `@ebay/skin/icon` can remain and continue to bundle all icons, but as mentioned it's size is going to significantly increase.

**Comment by ianmcburnie** [comment]
Can potentially wait till after v6. Doesn't seem urgent for coreui.

**Comment by ianmcburnie** [comment]
Current approach is looking like this:

Before (all in one dist file), ~86kb:

```css
svg.icon--arrow-left {
  height: 19px;
  width: 21px;
}
span.icon--arrow-left {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMSIgaGVpZ2h0PSIyMCI+PHBhdGggZD0iTTQuNjgzIDguNTY3bDYuOTAyLTYuNTk2Yy40ODgtLjQ2OC40My0xLjIzOCAwLTEuNjUtLjQzMi0uNDEyLTEuMjY5LS40NDMtMS43MjUgMEM5LjQwMi43NjQgMCA5Ljc0NSAwIDkuNzQ1czkuMjkzIDguODU0IDkuODYgOS40MjRjLjU2Ni41NyAxLjI4NS40MjIgMS43MjUgMCAuNDQxLS40MjEuNTU4LTEuMDkyIDAtMS42NDgtLjU1OC0uNTU3LTYuOTAyLTYuNTk3LTYuOTAyLTYuNTk3SDE5LjcyYy44NjYgMCAxLjI4LS41ODkgMS4yOC0xLjE4IDAtLjU4OC0uNDEtMS4xNzctMS4yOC0xLjE3N0g0LjY4M3oiLz48L3N2Zz4=');
  height: 19px;
  width: 21px;
}
svg.icon--arrow-right {
  height: 19px;
  width: 21px;
}
span.icon--arrow-right {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMSIgaGVpZ2h0PSIyMCI+PHBhdGggZD0iTTE2LjMxNyAxMC45NzZzLTYuNDEzIDYuMTI5LTYuOSA2LjU5N2MtLjQ5LjQ3LS40MzIgMS4yMzYgMCAxLjY1LjQzLjQxMiAxLjI2Ny40NDMgMS43MjMgMEMxMS41OTggMTguNzggMjEgOS43OTggMjEgOS43OThTMTEuNzA4Ljk0MyAxMS4xNC4zNzNjLS41NjYtLjU2OC0xLjI4My0uNDIxLTEuNzIzIDAtLjQ0Mi40MjItLjU2IDEuMDkzIDAgMS42NS41NTYuNTU2IDYuOSA2LjU5NyA2LjkgNi41OTdIMS4yODJDLjQxNCA4LjYyIDAgOS4yMDggMCA5Ljc5OHMuNDEgMS4xNzggMS4yODIgMS4xNzhoMTUuMDM1eiIvPjwvc3ZnPg==');
  height: 19px;
  width: 21px;
}
etc etc
```

After, removing base64 styles, ~6KB:

```css
span.icon--arrow-left,
svg.icon--arrow-left {
  height: 19px;
  width: 21px;
}
span.icon--arrow-right,
svg.icon--arrow-right {
  height: 19px;
  width: 21px;
}
```

Proposed, a new `icon-background.css` dist file ~80KB:

```css
span.icon--arrow-left {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMSIgaGVpZ2h0PSIyMCI+PHBhdGggZD0iTTQuNjgzIDguNTY3bDYuOTAyLTYuNTk2Yy40ODgtLjQ2OC40My0xLjIzOCAwLTEuNjUtLjQzMi0uNDEyLTEuMjY5LS40NDMtMS43MjUgMEM5LjQwMi43NjQgMCA5Ljc0NSAwIDkuNzQ1czkuMjkzIDguODU0IDkuODYgOS40MjRjLjU2Ni41NyAxLjI4NS40MjIgMS43MjUgMCAuNDQxLS40MjEuNTU4LTEuMDkyIDAtMS42NDgtLjU1OC0uNTU3LTYuOTAyLTYuNTk3LTYuOTAyLTYuNTk3SDE5LjcyYy44NjYgMCAxLjI4LS41ODkgMS4yOC0xLjE4IDAtLjU4OC0uNDEtMS4xNzctMS4yOC0xLjE3N0g0LjY4M3oiLz48L3N2Zz4=');
}
span.icon--arrow-right {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMSIgaGVpZ2h0PSIyMCI+PHBhdGggZD0iTTE2LjMxNyAxMC45NzZzLTYuNDEzIDYuMTI5LTYuOSA2LjU5N2MtLjQ5LjQ3LS40MzIgMS4yMzYgMCAxLjY1LjQzLjQxMiAxLjI2Ny40NDMgMS43MjMgMEMxMS41OTggMTguNzggMjEgOS43OTggMjEgOS43OThTMTEuNzA4Ljk0MyAxMS4xNC4zNzNjLS41NjYtLjU2OC0xLjI4My0uNDIxLTEuNzIzIDAtLjQ0Mi40MjItLjU2IDEuMDkzIDAgMS42NS41NTYuNTU2IDYuOSA2LjU5NyA2LjkgNi41OTdIMS4yODJDLjQxNCA4LjYyIDAgOS4yMDggMCA5Ljc5OHMuNDEgMS4xNzggMS4yODIgMS4xNzhoMTUuMDM1eiIvPjwvc3ZnPg==');
}
```

So, to stay backwards compatible, we could quickly release this in 6.3.0 and add a new Lasso flag that conditionally opts out of bringing in the huge `icon-background.css` file. In v7 we could flip this to make it an opt in.

For v7 we can also potentially strip the base file right down to this with no grouped selectors:

```css
.icon--arrow-left {
  height: 19px;
  width: 21px;
}
.icon--arrow-right {
  height: 19px;
  width: 21px;
}
```

But right now we need the specificity due to the way the base icon mixin works. We can refactor in v7.

**Comment by senthilp** [comment]
Sounds like a good plan :thumbsup:

**Comment by ianmcburnie** [comment]
I guess the browser.json would end up looking like this for now, but not sure this is going to help the ebayUI situation:

```json
{
    "dependencies": [
        {
            "if-not-flag": "skin-ds6",
            "dependencies": [
                "@ebay/skin/iconfont",
                "./dist/icon/ds4/icon.css"
            ]
        },
        {
            "path": "./dist/icon/ds4/icon-background.css",
            "if": "!flags.contains('skin-ds6') && !flags.contains('skin-bg-icon-opt-out')"
        },
        {
            "path": "./dist/icon/ds6/icon.css",
            "if-flag": "skin-ds6"
        },
        {
            "path": "./dist/icon/ds6/icon-background.css",
            "if": "flags.contains('skin-ds6') && !flags.contains('skin-bg-icon-opt-out')"
        }
    ]
}
```

Perhaps a full split into two modules, like `@ebay/skin/icon` & `@ebay/skin/base64` would be needed. I'll check with @DylanPiercey

**Comment by ianmcburnie** [comment]
Talked with Dylan offline, and splitting into two Lasso modules seems the way to go.

**Comment by ianmcburnie** [comment]
Ok, this should work. A slight refinement on what I posted previously:

dist/ds6/icon.css (contains no base64):

```css
span.icon--arrow-left,
svg.icon--arrow-left {
  height: 19px;
  width: 21px;
}
span.icon--arrow-right,
svg.icon--arrow-right {
  height: 19px;
  width: 21px;
}
etc etc
```

dist/ds6/base64.css (a new module):

```css
span.icon--arrow-left {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMSIgaGVpZ2h0PSIyMCI+PHBhdGggZD0iTTQuNjgzIDguNTY3bDYuOTAyLTYuNTk2Yy40ODgtLjQ2OC40My0xLjIzOCAwLTEuNjUtLjQzMi0uNDEyLTEuMjY5LS40NDMtMS43MjUgMEM5LjQwMi43NjQgMCA5Ljc0NSAwIDkuNzQ1czkuMjkzIDguODU0IDkuODYgOS40MjRjLjU2Ni41NyAxLjI4NS40MjIgMS43MjUgMCAuNDQxLS40MjEuNTU4LTEuMDkyIDAtMS42NDgtLjU1OC0uNTU3LTYuOTAyLTYuNTk3LTYuOTAyLTYuNTk3SDE5LjcyYy44NjYgMCAxLjI4LS41ODkgMS4yOC0xLjE4IDAtLjU4OC0uNDEtMS4xNzctMS4yOC0xLjE3N0g0LjY4M3oiLz48L3N2Zz4=');
}
span.icon--arrow-right {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMSIgaGVpZ2h0PSIyMCI+PHBhdGggZD0iTTE2LjMxNyAxMC45NzZzLTYuNDEzIDYuMTI5LTYuOSA2LjU5N2MtLjQ5LjQ3LS40MzIgMS4yMzYgMCAxLjY1LjQzLjQxMiAxLjI2Ny40NDMgMS43MjMgMEMxMS41OTggMTguNzggMjEgOS43OTggMjEgOS43OThTMTEuNzA4Ljk0MyAxMS4xNC4zNzNjLS41NjYtLjU2OC0xLjI4My0uNDIxLTEuNzIzIDAtLjQ0Mi40MjItLjU2IDEuMDkzIDAgMS42NS41NTYuNTU2IDYuOSA2LjU5NyA2LjkgNi41OTdIMS4yODJDLjQxNCA4LjYyIDAgOS4yMDggMCA5Ljc5OHMuNDEgMS4xNzggMS4yODIgMS4xNzhoMTUuMDM1eiIvPjwvc3ZnPg==');
}
etc etc
```

base64.browser.json:

```json
{
    "dependencies": [
        {
            "path": "./dist/base64/ds4/base64.css",
            "if-not-flag": "skin-ds6"
        },
        {
            "path": "./dist/base64/ds6/base64.css",
            "if-flag": "skin-ds6"
        }
    ]
}
```

icon.browser.json (notice the new base64 dependency):

```json
{
    "dependencies": [
        "@ebay/skin/base64",
        {
            "if-not-flag": "skin-ds6",
            "path": "./dist/icon/ds4/icon.css"
        },
        {
            "if-flag": "skin-ds6",
            "path": "./dist/icon/ds6/icon.css"
        }
    ]
}
```

For 6.3.0, we need the base64 always included by default to keep backwards compatibility (but I could make it opt out with a Lasso flag). This would be an interim step to get we want to.

For skin 7.0.0 I could wrap the  "@ebay/skin/base64" inside of a conditional to make it opt-in only, if we prefer.

**Comment by seangates** [comment]
This looks good, as long as we still plan to add the opt-out flag in another step. I assume we'll do that as a further refinement of #457? Or is this deemed part of this splitting work?

Could/should we then remove the `@ebay/skin/icon` from `@ebay/skin/combo` altogether (sans any opt-out) since it's being pulled in when needed by the components?

**Comment by DylanPiercey** [comment]
@ianmcburnie that sounds reasonable to me.

**Comment by ianmcburnie** [comment]
Fix merged to 6.3.0 branch. Closing ticket.
